test/extended/router/gatewayapicontroller.go

```diff
@@ -673,3 +768,224 @@ func isOKD(oc *exutil.CLI) (bool, error) {
+			e2e.Logf("Trying deleting the original gateway lb service %q, retrying...", gatewayLbService)
+			return false, nil
+		}
+
+		if len(lbService.Status.LoadBalancer.Ingress) == 0 {
+			e2e.Logf("New gateway lb service %q is created, but without lb hostname or ip, retrying...", gatewayLbService)
+			return false, nil
+		}
+
+		if lbService.Status.LoadBalancer.Ingress[0].Hostname != "" {
+			lbAddress = lbService.Status.LoadBalancer.Ingress[0].Hostname
+		} else {
+			lbAddress = lbService.Status.LoadBalancer.Ingress[0].IP
+		}
+		if lbAddress == "" {
+			e2e.Logf("No load balancer address for service %q, retrying", gatewayLbService)
+			return false, nil
+		}
+
+		e2e.Logf("Got load balancer address for service %q: %v", gatewayLbService, lbAddress)
+		return true, nil
+	})
+	o.Expect(err).NotTo(o.HaveOccurred())
+}
+
+// used to get a gateway dnsrecord from a given dnsrecordList
+func getGWDNSRecords(dnsrecordList *operatoringressv1.DNSRecordList, gwName string) (operatoringressv1.DNSRecord, error) {
+	for _, dnsrecord := range dnsrecordList.Items {
+		if strings.Contains(dnsrecord.ObjectMeta.Name, gwName+"-") {
+			return dnsrecord, nil
+		}
+	}
+
+	return operatoringressv1.DNSRecord{}, errors.New("Could not get the name of the gw dnsrecord")
+}
+
+// used to wait for the gateway dnsrecord is automatically recreated successfully
+func pollWaitGWDNSRecordsRecreated(oc *exutil.CLI, gwName, ingressNamespace, expectedtargets string, originalCreatedTime metav1.Time) {
+	err := wait.Poll(3*time.Second, 300*time.Second, func() (bool, error) {
+		dnsrecordList, err := oc.AdminIngressClient().IngressV1().DNSRecords(ingressNamespace).List(context.Background(), metav1.ListOptions{})
+		if err != nil {
+			e2e.Logf("Failed to List DNSRecords in namespace %q: %v, retrying...", ingressNamespace, err)
+			return false, nil
+		}
+
+		dnsrecord, err := getGWDNSRecords(dnsrecordList, gwName)
+		if err != nil {
+			e2e.Logf("Failed to get the DNSRecord name for gateway %q in namespace %q: %v, retrying...", gwName, ingressNamespace, err)
+			return false, nil
+		}
+
+		if dnsrecord.ObjectMeta.CreationTimestamp == originalCreatedTime {
+			e2e.Logf("Trying deleting the original dnsrecord of GW %q, retrying...", gwName)
+			return false, nil
+		}
+
+		currentTargets := getSortedString(dnsrecord.Spec.Targets)
+		if currentTargets != expectedtargets {
+			e2e.Logf("Current DNSRecord targets %q for gateway %q differ from expected %q, retrying...", currentTargets, gwName, expectedtargets)
+			return false, nil
+		}
+
+		for _, zone := range dnsrecord.Status.Zones {
+			for _, condition := range zone.Conditions {
+				if condition.Status != "True" || condition.Reason != "ProviderSuccess" {
+					e2e.Logf(`Current condition Status(%q) and Current condition Reason(%q) of zone %q, at least one of them is not matched, retrying...`, condition.Status, condition.Reason, zone)
+					return false, nil
+				}
+			}
```

Comment:
I think that this condition is buggy. We try to iterate over all conditions but we may stop at the first one. Because if the first one's reason is not `ProviderSuccess` and its status is not `True` we retry. We should do [the same check](https://github.com/openshift/origin/blob/77a751eb8cf4b4dfba4ae4695cb11aba9bb8c34b/test/extended/router/gatewayapicontroller.go#L433-L439) as in `assertDNSRecordStatus`: look for `Published==True` condition.

Reply:
Thanks for the review. From my side, there may be quite a few zones, and there may be quite a few conditions under a zone, only all conditions are ok, we can say the gw dnsrecord is good.
In [the same check](https://github.com/openshift/origin/blob/77a751eb8cf4b4dfba4ae4695cb11aba9bb8c34b/test/extended/router/gatewayapicontroller.go#L433-L439), only one condition is ok, we can say the gw dnsrecord is good.

We just need to make sure one condition of a dnsrecord's zone is in good status, right?

Reply:
The DNSRecord's status is generic and assumes that publishing can be done in multiple zones. Practically, I don't think I ever saw more than 1 zone. That's why the existing check from `assertDNSRecordStatus` works fine even if it returns at first `Published=True` condition found. I think something like this should do fine in your case:
```
if condition.Type == "Published" && condition.Status != "True" {
    e2e.Logf(`DNSRecord %q is not published in zone %q, retrying...`, dnsRecord.Name, zone)
    return false, nil
}
```
This won't change your approach a lot and will make sure the record is published on all zones.

Reply:
Yes, currently it is only one zone. I will update it, thanks.